jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/proxy/WebSocketProxyTest.java

```diff
@@ -0,0 +1,379 @@
+
+        closeStatus = CloseStatus.getCloseStatus(serverFrameHandler.receivedFrames.poll());
+        assertThat(closeStatus.getCode(), is(CloseStatus.SERVER_ERROR));
+        assertThat(closeStatus.getReason(), containsString("simulated client onOpen error"));
+
+        assertNull(clientHandler.receivedFrames.poll());
+    }
+
+
+
+    @Test
+    public void testServerError() throws Exception
+    {
+        serverFrameHandler.throwOnFrame();
+        WebSocketProxy.Client2Proxy proxyClientSide = proxy.client2Proxy;
+        WebSocketProxy.Server2Proxy proxyServerSide = proxy.server2Proxy;
+
+        BasicFrameHandler clientHandler = new BasicFrameHandler("CLIENT");
+        ClientUpgradeRequest upgradeRequest = ClientUpgradeRequest.from(_client, new URI("ws://localhost:8080/proxy/test"), clientHandler);
+
+        CompletableFuture<CoreSession> response = _client.connect(upgradeRequest);
+        response.get(5, TimeUnit.SECONDS);
+        clientHandler.sendText("hello world");
+        clientHandler.awaitClose();
+        serverFrameHandler.awaitClose();
+        awaitProxyClose(proxyClientSide, proxyServerSide);
+
+        CloseStatus closeStatus;
+        Frame frame;
+
+        // Client
+        frame = clientHandler.receivedFrames.poll();
+        assertThat(CloseStatus.getCloseStatus(frame).getCode(), is(CloseStatus.SERVER_ERROR));
+
+        // Client2Proxy
+        frame = proxyClientSide.receivedFrames.poll();
+        assertThat(frame.getOpCode(), is(OpCode.TEXT));
+        assertThat(frame.getPayloadAsUTF8(), is("hello world"));
+
+        frame = proxyClientSide.receivedFrames.poll();
+        closeStatus = CloseStatus.getCloseStatus(frame);
+        assertThat(closeStatus.getCode(), is(CloseStatus.SERVER_ERROR));
+        assertThat(closeStatus.getReason(), is("intentionally throwing in server onFrame()"));
+
+        frame = proxyClientSide.receivedFrames.poll();
+        assertNull(frame);
+        assertThat(proxyClientSide.getState(), is(WebSocketProxy.State.CLOSED));
+
+        // Server2Proxy
+        frame = proxyServerSide.receivedFrames.poll();
+        closeStatus = CloseStatus.getCloseStatus(frame);
+        assertThat(closeStatus.getCode(), is(CloseStatus.SERVER_ERROR));
+        assertThat(closeStatus.getReason(), is("intentionally throwing in server onFrame()"));
+
+        frame = proxyServerSide.receivedFrames.poll();
+        assertNull(frame);
+        assertThat(proxyServerSide.getState(), is(WebSocketProxy.State.CLOSED));
+
+        // Server
+        frame = serverFrameHandler.receivedFrames.poll();
+        assertThat(frame.getOpCode(), is(OpCode.TEXT));
+        assertThat(frame.getPayloadAsUTF8(), is("hello world"));
+        frame = serverFrameHandler.receivedFrames.poll();
+        assertNull(frame);
+    }
+
+    @Test
+    public void testServerErrorClientNoResponse() throws Exception
+    {
+        serverFrameHandler.throwOnFrame();
```

Comment:
I think this should be `serverFrameHandler.noResponseOnFrame()`?

Reply:
this test was to test if what would happen in the Proxy if the server had an error and the client didn't respond to the close frame it was sent, so the client handler is overridden to not complete the close frame callback